packages/logging/lokiLogProvider.go

```diff
@@ -0,0 +1,274 @@
+package logging
+
+import (
+	"cmp"
+	"encoding/json"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"net/url"
+	"slices"
+	"strconv"
+	"time"
+)
+
+const defaultQueryLimit = 1000
+
+type nowClock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (r *realClock) Now() time.Time {
+	return time.Now()
+}
+
+type LokiLogProvider struct {
+	gatewayUrl string
+	username   string
+	password   string
+	clock      nowClock
+}
+
+func NewLokiLogProvider(gatewayUrl string, username string, password string) *LokiLogProvider {
+	return &LokiLogProvider{
+		gatewayUrl: gatewayUrl,
+		username:   username,
+		password:   password,
+		clock:      &realClock{},
+	}
+}
+
+func (llp *LokiLogProvider) getLogs(doguName string, linesCount int) ([]logLine, error) {
+	query := fmt.Sprintf("{pod=~\"%s.*\"}", doguName)
+	endDate := llp.clock.Now()
+	startDate := createQueryStartDateFromEndDate(endDate)
+
+	result := make([]logLine, 0)
+	for {
+		limit := calculateQueryLimit(linesCount, len(result))
+
+		logrus.Debugf("running loki query for '%s' from %s to %s with limit %d", doguName, startDate.Format(time.RFC3339), endDate.Format(time.RFC3339), limit)
+		lokiQueryUrl, err := buildLokiQueryUrl(llp.gatewayUrl, query, startDate, endDate, limit)
+		if err != nil {
+			return result, fmt.Errorf("failed to build loki-query: %v", err)
+		}
+
+		lokiResponse, err := llp.doLokiHttpQuery(lokiQueryUrl)
+		if err != nil {
+			return result, fmt.Errorf("faild to execute loki-query: %v", err)
+		}
+
+		logLines, err := extractLogLinesFromLokiResponse(lokiResponse)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract logs from loki response: %v", err)
+		}
+
+		if len(logLines) <= 0 {
+			// no new lines to read => nothing is left
+			break
+		}
+
+		// prepend logs
+		result = append(logLines, result...)
+
+		if len(logLines) < limit {
+			// the query returned fewer lines than requested => nothing is left
+			break
+		}
+
+		if linesCount > 0 && len(result) >= linesCount {
+			// we reached the maximum lines count
+			break
+		}
+
+		// there are still logs to read -> start with the newest log timestamp from the last response
+		endDate = logLines[0].timestamp
+		startDate = createQueryStartDateFromEndDate(endDate)
+	}
+
+	// because multiple logs can happen at the exact same timestamp and the query is batched over time,
+	// it is possible that consecutive batches contain the same logLines. These need to be removed.
+	result = deduplicateLogLines(result)
```

Comment:
```suggestion
		endDate = logLines[0].timestamp.Add(-1 * time.Nanosecond)
		startDate = createQueryStartDateFromEndDate(endDate)
	}
```
Wouldn't this get rid of the need to deduplicate log lines?

Reply:
let`s say that 5 logs happen at the exact same time. In the first batch the result only contains 3 of those 5 logs.
With this method we would loose the other 2 logs

Reply:
Yeah, I didn't understand that we request a limited number of logs from the server.